compiler/rustc_session/src/options.rs

```diff
@@ -1457,6 +1471,9 @@ options! {
         "print layout information for each type encountered (default: no)"),
     proc_macro_backtrace: bool = (false, parse_bool, [UNTRACKED],
          "show backtraces for panics during proc-macro execution (default: no)"),
+    proc_macro_execution_strategy: ProcMacroExecutionStrategy = (ProcMacroExecutionStrategy::SameThread,
+        parse_proc_macro_execution_strategy, [UNTRACKED],
+        "how to run proc-macro code (default: same-thread)"),
```

Comment:
@pnkfelix I'm running out of ideas... what if you move this to the end? (as in, maybe the order of the fields or their offsets, in this huge `struct`, is causing weird things)

It could even be its presence - the flag could be removed and `exec_strategy` (the sole user of the flag, in `compiler/rustc_expand/src/proc_macro.rs`) made to, idk, use `std::env::var` instead? (or just hardcode the choice, but that might cause changes in optimizations)

Reply:
Wacky! I moved the field to the end of the struct, and that made the two methods with the two highest instruction counts (`parse_tt` and `MacroRulesMacroExpander::expand`) each have lower instruction-counts (Ir) according to `cachegrind`, but the overall instruction-count increased:

Instruction Counts | [8f68c43c](https://github.com/rust-lang/rust/commit/8f68c43ca6a6381a4d73f887f112e9fb95769905) | [bd84c73f](https://github.com/rust-lang/rust/commit/bd84c73ffe0a54ce2d77c92948a26ffa8fec04a3) | field-at-end
--------------------|-------------------------|---------------------|--------
TOTAL                     | 988,908,196 (100.0%)    | 1,002,719,736 (100.0%) | 1,138,908,984 (100.0%)
parse_tt                  | 134,903,906 (13.64%)    | 149,152,914 (14.87%)  | 99,794,157 (8.76%)
MacroRules<br/>MacroExpander<br/>::expand |  56,387,220 (5.70%) |  56,379,944 (5.62%) | 50,969,953 (4.48%)
SipHasher128::finish128 | 5,975,227 (0.60%) | 5,975,227 (0.60%) | 36,508,882 (3.21%)
token_name_eq | (not in report) | (not in report) | 29,133,905 (2.56%)
try_mark_previous_green | 53,862,272 (5.45%) |  53,872,906 (5.37%) | 25,344,946 (2.23%)

I don't really want to spend too much more trying to dissect this thing that is probably just an LLVM codegen oddity, but I will at least try eliminating the field entirely as you suggest.

(As an aside, the fact that this kind of field shuffling had such an enormous impact on `parse_tt` does make me wonder if there's some kind of low-hanging fruit embedded in the code there, e.g. maybe the code is accessing the session when it should be locally stashing an unchanging value? But isn't that the kind of thing you'd like to believe your compiler is doing *for you* automatically? 😆 )

Reply:
I suspect that the code is hot, and also that register spilling is happening in a different way or inlining choices are a little different. I.e. the kind of thing that's hard to control, and even if you manage to get it in the faster state now a tiny change later on might make it go back to the slower state.